build.rs

```diff
@@ -23,28 +23,62 @@
 
 extern crate gcc;
 
+use std::env;
+use std::path::PathBuf;
+
+const ANDROID_AARCH64_COMPILER: &'static str = "toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/bin";
+const ANDROID_ARM_COMPILER: &'static str = "toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin";
+const ANDROID_I686_COMPILER: &'static str = "toolchains/x86-4.9/prebuilt/darwin-x86_64/bin";
+const ANDROID_INCLUDE: &'static str = "platforms/android-21/arch-arm64/usr/include";
+
+fn concat_paths(first: &str, second: &str) -> PathBuf {
+	let mut path = PathBuf::from(first);
+	path.push(second);
+	path
+}
+
+fn setup_android(config: &mut gcc::Config) {
+	let path = env::var_os("PATH").unwrap();
+	let ndk_home = env::var("NDK_HOME").expect("NDK_HOME is not set");
+	let mut paths = env::split_paths(&path).collect::<Vec<_>>();
+	paths.push(concat_paths(&ndk_home, ANDROID_AARCH64_COMPILER));
+	paths.push(concat_paths(&ndk_home, ANDROID_ARM_COMPILER));
+	paths.push(concat_paths(&ndk_home, ANDROID_I686_COMPILER));
+
+	let new_path = env::join_paths(paths).unwrap();
+	env::set_var("PATH", new_path);
+
+	config.include(&concat_paths(&ndk_home, ANDROID_INCLUDE));
+}
+
 fn main() {
-    let mut base_config = gcc::Config::new();
-    base_config.include("depend/secp256k1/")
-               .include("depend/secp256k1/include")
-               .include("depend/secp256k1/src")
-               .flag("-g")
-               // TODO these three should be changed to use libgmp, at least until secp PR 290 is merged
-               .define("USE_NUM_NONE", Some("1"))
-               .define("USE_FIELD_INV_BUILTIN", Some("1"))
-               .define("USE_SCALAR_INV_BUILTIN", Some("1"))
-               // TODO these should use 64-bit variants on 64-bit systems
-               .define("USE_FIELD_10X26", Some("1"))
-               .define("USE_SCALAR_8X32", Some("1"))
-               .define("USE_ENDOMORPHISM", Some("1"))
-               // These all are OK.
-               .define("ENABLE_MODULE_ECDH", Some("1"))
-               .define("ENABLE_MODULE_SCHNORR", Some("1"))
-               .define("ENABLE_MODULE_RECOVERY", Some("1"));
-
-    // secp256k1
-    base_config.file("depend/secp256k1/contrib/lax_der_parsing.c")
-	       .file("depend/secp256k1/src/ext.c")
-               .compile("libsecp256k1.a");
+	let mut base_config = gcc::Config::new();
+	base_config.include("depend/secp256k1/")
+		.include("depend/secp256k1/include")
+		.include("depend/secp256k1/src");
+
+	let target = env::var("TARGET").unwrap();
```

Comment:
it's always set by `cargo` [docs](http://doc.crates.io/environment-variables.html)